<html><head></head><body>
<div><h1 class="chapter-number" id="_idParaDest-145"><a id="_idTextAnchor145" class="pcalibre calibre6 pcalibre1"/>11</h1>
<h1 id="_idParaDest-146" class="calibre5"><a id="_idTextAnchor146" class="pcalibre calibre6 pcalibre1"/>Mapping Language Families – an Ontological Approach</h1>
<p class="calibre3">In this chapter, we’ll explore an important area of science called linguistics. <strong class="bold">Linguistics</strong> is the study<a id="_idIndexMarker498" class="pcalibre calibre6 pcalibre1"/> of languages, including word usage patterns, grammar evolution, and the social constructs of language. We’ll <a id="_idIndexMarker499" class="pcalibre calibre6 pcalibre1"/>explore <strong class="bold">ontologies</strong>, which relate concepts such as words to each other through defined relationships, and <a id="_idIndexMarker500" class="pcalibre calibre6 pcalibre1"/>build <strong class="bold">language families</strong>, where languages within a family are related to each other across time and geography.</p>
<p class="calibre3">Network science provides us with tools to summarize and compare ontologies and language families to study hypothesized similarities and differences. We’ll quantify differences in hypothesized language families at the end of this chapter, though the methodology can apply to ontologies or any other network-based structures derived from linguistic studies.</p>
<p class="calibre3">In this chapter, we will cover the following topics:</p>
<ul class="calibre10">
<li class="calibre11">What is an ontology?</li>
<li class="calibre11">Language families</li>
<li class="calibre11">Mapping language families</li>
</ul>
<p class="calibre3">By the end of this chapter, you’ll know some of the basic tools used in linguistic-related analyses and be able to compare different networks to understand the key differences between them. Because these tools scale well, you’ll be able to tackle even very large ontologies.</p>
<h1 id="_idParaDest-147" class="calibre5"><a id="_idTextAnchor147" class="pcalibre calibre6 pcalibre1"/>Technical requirements</h1>
<p class="calibre3">You will require Jupyter Notebook to run the practical examples in this chapter. </p>
<p class="calibre3">The code for this chapter is available here: <a href="https://github.com/PacktPublishing/Modern-Graph-Theory-Algorithms-with-Python" class="pcalibre calibre6 pcalibre1">https://github.com/PacktPublishing/Modern-Graph-Theory-Algorithms-with-Python</a></p>
<h1 id="_idParaDest-148" class="calibre5"><a id="_idTextAnchor148" class="pcalibre calibre6 pcalibre1"/>What is an ontology?</h1>
<p class="calibre3">In this section, we’ll <a id="_idIndexMarker501" class="pcalibre calibre6 pcalibre1"/>learn about ontologies and how they can be represented and stored as networks. While this chapter focuses on linguistic ontologies, many other types of ontologies exist, including gene ontologies, briefly reviewed to motivate this section.</p>
<h2 id="_idParaDest-149" class="calibre7"><a id="_idTextAnchor149" class="pcalibre calibre6 pcalibre1"/>Introduction to ontologies</h2>
<p class="calibre3">As mentioned in the chapter introduction, an ontology is a set of ideas that are related by some property. Generally, they are used to organize knowledge within a specific discipline. Let’s make this a little bit more concrete with an example. Because examples in genetics are more widely used in ontology study, we’ll build up the intuition with a biological example before tackling linguistics.</p>
<p class="calibre3">Say we have five genes that are within a regulatory pathway that controls the production of enzymes or polypeptides found in an animal’s venom. Often, environmental conditions and geography play a role in the ratio of active compounds found in venom, even within the same species, through regulating gene transcription within the venom glands. Before we start collecting information on these genes and their influence on venom production, we simply have a collection of genes, as shown in <em class="italic">Figure 11</em><em class="italic">.1</em>:</p>
<div><div><img alt="Figure 11.1 – A collection of five genes involved in regulating venom production" src="img/B21087_11_01.jpg" class="calibre4"/>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US">Figure 11.1 – A collection of five genes involved in regulating venom production</p>
<p class="calibre3">To better understand the genes in <em class="italic">Figure 11</em><em class="italic">.1</em>, we might run an environmental experiment with varying conditions to turn genes on and off and elucidate their relationships with each other—the dependencies that exist within this regulatory pathway. Interdependencies can be represented by connecting each of these gene vertices with an edge. Let’s say we ran several experiments and found the following relationships:</p>
<div><div><img alt="Figure 11.2 – An ontology representing relationships between genes that regulate venom production" src="img/B21087_11_02.jpg" class="calibre4"/>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US">Figure 11.2 – An ontology representing relationships between genes that regulate venom production</p>
<p class="calibre3"><em class="italic">Figure 11</em><em class="italic">.2</em> shows<a id="_idIndexMarker502" class="pcalibre calibre6 pcalibre1"/> that all four other genes are downstream from <em class="italic">Gene 1</em>, suggesting that this gene is critically important to the regulation of a critical compound in the animal’s venom. Mutations in <em class="italic">Gene 1</em> likely result in loss of function (and potentially endanger the animal). Visually representing this ontology as a network allows us to easily pinpoint key information within the system we are studying.</p>
<p class="calibre3">However, many organisms’ gene regulation pathways are much larger than this example and are hard to visualize in practice—particularly entire gene ontologies, where the organisms’ full set of genes is mapped and connected via functional pathways. This means that we’ll need to use network analytics tools to study differences in proposed gene regulation pathways or compare gene regulation pathways across organisms that show the same function. For instance, in our venom example, we might be interested in mapping the full gene ontology of a species of sea krait, a member of the elapid family, and comparing its full gene ontology with another elapid species, such as taipan or a black mamba. These species can have upward of 12,000 genes, creating a very large ontology comparison problem.</p>
<p class="calibre3">Let’s return to our focus on linguistics and see how language-related ontologies are constructed and can be used to study language differences among dialects, pidgins, or socioeconomic groups—or simply to understand the structure and function of a language.</p>
<p class="calibre3">Consider two words in the English language: <em class="italic">go</em> and <em class="italic">went</em>. Both words are verbs. Both have the same semantic meaning with respect to the sentence’s subject. However, one is present tense and the other is past tense. All these pieces of information help us understand the functionality of these words within the English language. We can map these relationships to<a id="_idIndexMarker503" class="pcalibre calibre6 pcalibre1"/> a network structure for each word, as <em class="italic">Figure </em><em class="italic">11</em><em class="italic">.3</em> summarizes:</p>
<div><div><img alt="Figure 11.3 – Two networks showing the linguistic properties of the words “go” and “went”" src="img/B21087_11_03.jpg" class="calibre4"/>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US">Figure 11.3 – Two networks showing the linguistic properties of the words “go” and “went”</p>
<p class="calibre3"><em class="italic">Figure 11</em><em class="italic">.3</em> summarizes the linguistic properties of our two words: <em class="italic">go</em> and <em class="italic">went</em>. Imagine creating networks like this for every word in the English language. We would have very many networks mapping linguistic properties to words, and it would be easy to lose important networks. We’d also use a lot of computing power trying to search through graph<a id="_idIndexMarker504" class="pcalibre calibre6 pcalibre1"/> collections for words or properties of interest. Let’s explore ontology creation further.</p>
<h2 id="_idParaDest-150" class="calibre7"><a id="_idTextAnchor150" class="pcalibre calibre6 pcalibre1"/>Representing information as an ontology</h2>
<p class="calibre3">To create an <a id="_idIndexMarker505" class="pcalibre calibre6 pcalibre1"/>ontology, we need to define concepts/words/terms of interest and the properties that they might share. In our <em class="italic">Figure 11</em><em class="italic">.3</em> example, we have a set of words with linguistic properties of interest. Once we have these two sets defined (words and linguistic properties, respectively, in <em class="italic">Figure 11</em><em class="italic">.3</em>’s example), we can create maps between these two sets (the lines shown in <em class="italic">Figure 11</em><em class="italic">.3</em>). Our word example was a very small example, though. Many times, these sets can include thousands or tens of thousands of items in each of our sets.</p>
<p class="calibre3">To deal with these larger sets of information and the potential relationships that exist between items in them, we need a systematic way to store both sets and their relationships. We’ll also need a lot of people to create this very large ontology, preferably with a way to share the information among all of the people working to create this ontology. That’s where databases come into play.</p>
<p class="calibre3"><strong class="bold">Ontology Web Language</strong> (<strong class="bold">OWL</strong>) database systems organize knowledge networks like the ones <a id="_idIndexMarker506" class="pcalibre calibre6 pcalibre1"/>shown in <em class="italic">Figure 11</em><em class="italic">.3</em>. Each word, term, or concept occupies a row within the database that is mapped to whichever properties of interest within the ontology match that word/term/concept, thereby creating a sort of database of databases to store word/term/concept data along with their relationships to key properties of interest in the ontology. From this database, we can search for information, visualize relationships of interest, and mine the entire ontology for overarching themes.</p>
<p class="calibre3">Any type of knowledge can be organized in an OWL database. For instance, we could use an OWL database to organize knowledge about animal taxonomies or evolutionary genetics to quickly search for relationships among colubrid snake species according to genetic analysis results. We could also organize medical terms to create an easy-to-use lookup system for doctors to use when confronted with symptoms or diseases beyond their areas of expertise to facilitate the diagnosis of a patient. We could even organize knowledge about character relationships in long fictional works to provide study guides for a book such as <em class="italic">War </em><em class="italic">and Peace</em>.</p>
<p class="calibre3">Now that we know the basics about ontologies and the systems in which we can store them, let’s turn our attention back to linguistics, where we may wish to study the evolution of <a id="_idIndexMarker507" class="pcalibre calibre6 pcalibre1"/>languages across language families.</p>
<h1 id="_idParaDest-151" class="calibre5"><a id="_idTextAnchor151" class="pcalibre calibre6 pcalibre1"/>Language families</h1>
<p class="calibre3"><strong class="bold">Language families</strong> are groups<a id="_idIndexMarker508" class="pcalibre calibre6 pcalibre1"/> of languages with a common ancestor that evolved (usually) within a related geographic region. For example, the Romance language family consists of languages that evolved from Vulgar Latin (late-period Latin) in the Southern European region. The Nilo-Saharan language family evolved in the Nile Delta during the time of the Nubian kingdom. There are roughly 150 language families rolled into major language families that exist today, plus several <strong class="bold">language isolates</strong> that <a id="_idIndexMarker509" class="pcalibre calibre6 pcalibre1"/>do not have relationships with other languages (Australian Tiwi, for instance). Language isolates have no genetic ancestors linked to other languages. Campbell gives seven major language families (one per continent), though other researchers classify major language families into other groups of languages.</p>
<h2 id="_idParaDest-152" class="calibre7"><a id="_idTextAnchor152" class="pcalibre calibre6 pcalibre1"/>Language drift and relationships</h2>
<p class="calibre3">Language families<a id="_idIndexMarker510" class="pcalibre calibre6 pcalibre1"/> emerged <a id="_idIndexMarker511" class="pcalibre calibre6 pcalibre1"/>as different populations of genetic ancestor languages became geographically isolated enough for the ancestor language to evolve regionally. As languages evolve, words may change. Pronunciations may change. Grammar may evolve. In<a id="_idIndexMarker512" class="pcalibre calibre6 pcalibre1"/> many English <strong class="bold">pidgins</strong> (mixes of two or more languages, including English in this case), English has mixed with local languages to create a hybrid language that retains some of the grammar and phrasing of English alongside a local language. In Nigerian English pidgins, English mixes with local languages, such as Igbo, Hausa, or Yoruba. For instance, “<em class="italic">How are you?</em>” becomes “<em class="italic">How na?</em>” and “<em class="italic">Please come join me for a meal</em>” becomes “<em class="italic">Abeg come </em><em class="italic">chop food</em>.”</p>
<p class="calibre3">Over time, pidgin languages can evolve to the point of being a separate language altogether (a child of English, much the way Middle English evolved from Old English as other languages influenced Old English) or remain a pidgin of the parent language (English, in this case). When the language evolves, it becomes a child branch of the parent language. In the case of English, its widespread use in business and school systems suggests that many child languages will develop over time into separate languages with distinct diction and grammar. Given pidgins in Nigeria and their widespread usage, a Nigerian English child language is probably the most likely child language to develop in the next century or so.</p>
<p class="calibre3">A few driving factors tend to favor language evolution into child languages or pidgins. Commerce between different countries or regions that do not overlap linguistically necessitates means of communication. Pidgins often develop in marketplaces where many speakers of different languages must communicate with each other to buy and sell goods. Common words and phrases might be used over and over during a transaction (such as currency values or amounts). Tricky grammar or articles might be dropped to facilitate understanding between speakers of two different languages.</p>
<p class="calibre3">Empire expansion is another driver of language evolution. The Roman Empire spread Latin throughout much of Europe, influencing languages, particularly those closest to Rome, over centuries. Spain and Portugal are interesting cases, where both Roman influences in Latin and Moor influences in Arabic influenced local dialects. Over time, the Romance languages, including Spanish and Portuguese, evolved out of Latin as the Roman Empire’s power waned and other linguistic and cultural influences separated regions that were united in their use of Latin for commerce, schooling, and religion. In fact, the Roman Empire influenced the development of these languages so directly that the term <strong class="bold">Romance languages</strong> refers<a id="_idIndexMarker513" class="pcalibre calibre6 pcalibre1"/> to the Roman Empire!</p>
<p class="calibre3">Most English pidgins that exist today owe their existence to British colonialism, as do different dialects of English spoken across Great Britain, the United States of America, and Australia. Depending on how long English lasts as business’s lingua franca, we may see many child languages evolving from English in the next centuries.</p>
<p class="calibre3">A <strong class="bold">phylogeny tree</strong> can be used <a id="_idIndexMarker514" class="pcalibre calibre6 pcalibre1"/>to summarize the evolutionary relationships between ideas, languages, or species over time visually by showing parent and child relationships. This makes the study of evolutionary structure easier, as the visuals provide clues about hypothesized pathways of evolution and a network structure that is easy to analyze for information and<a id="_idIndexMarker515" class="pcalibre calibre6 pcalibre1"/> potential statistical<a id="_idIndexMarker516" class="pcalibre calibre6 pcalibre1"/> differences. General ontology structures do not have a hierarchical structure, where one part of the network is ideologically dependent on a prior part of the network. Let’s see this in action with the hypothesized evolutionary pathways of Nilo-Saharan languages.</p>
<h2 id="_idParaDest-153" class="calibre7"><a id="_idTextAnchor153" class="pcalibre calibre6 pcalibre1"/>Nilo-Saharan languages</h2>
<p class="calibre3">Nilo-Saharan <a id="_idIndexMarker517" class="pcalibre calibre6 pcalibre1"/>languages include languages spoken <a id="_idIndexMarker518" class="pcalibre calibre6 pcalibre1"/>in South Sudan, Ethiopia, Sudan, Kenya, Tanzania, Uganda, Chad, Burkina Faso, Cameroon, Mali, the Democratic Republic of Congo, Benin, Algeria, Nigeria, Niger, Egypt, and Libya. Further small language communities exist in several other countries, as well. Some common languages included in this family are Maasai, Songhay, Dinka, Luo, and Kanuri. About 70 million people speak a Nilo-Saharan language, with the number of speakers increasing for some regions and dialects.</p>
<p class="calibre3">The linguistic study of Nilo-Saharan languages dates to at least the 1800s, and many papers in the last decade have debated language drift and relationships of parent languages with their child languages. In this section, we’ll examine a few of the proposed hierarchies and then return to analyze the similarities and differences of these hierarchies with Python code in the next section. If you are interested in the papers themselves, you can consult the references at the end of the chapter, particularly the Dimmendaal paper, which summarizes recent debates on parent-child relationships in the Nilo-Saharan family.</p>
<p class="calibre3">Greenberg proposed the following hierarchy in 1963, classifying several language family branches. <em class="italic">Figure 11</em><em class="italic">.4</em> shows this hierarchy, including children branches off a child branch of the larger Nilo-Saharan family:</p>
<div><div><img alt="Figure 11.4 – Greenberg’s classification of Nilo-Saharan languages" src="img/B21087_11_04.jpg" class="calibre4"/>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US">Figure 11.4 – Greenberg’s classification of Nilo-Saharan languages</p>
<p class="calibre3"><em class="italic">Figure 11</em><em class="italic">.4</em> shows some subclassifications of Nilo-Saharan languages that are part of the language family. Each of these subclassifications continues to individual languages, totaling roughly 130 languages spoken in this region that originated near the area where the Nile tributaries meet in what was once Nubia.</p>
<p class="calibre3">Since 1963, many papers have further divided these subclasses and added additional subclasses culminating in the Dimmendaal paper that we’ll discuss in this section, occasionally creating very complicated and deep network structures. Grammar, diction, phonology (pronunciation of consonants, in this case), and terminology typically dictate where lines are drawn, but languages are often fluid and may overlap somewhat, making classification systems difficult to discern. Let’s turn to one of the more recent works on the Nilo-Saharan family of languages.</p>
<p class="calibre3">Dimmendaal’s recent <a id="_idIndexMarker519" class="pcalibre calibre6 pcalibre1"/>work proposes a new hierarchy<a id="_idIndexMarker520" class="pcalibre calibre6 pcalibre1"/> of the Nilo-Saharan subfamilies, summarized in <em class="italic">Figure 11</em><em class="italic">.5</em>:</p>
<div><div><img alt="Figure 11.5 – Dimmendaal’s hierarchy of Nilo-Saharan subfamilies" src="img/B21087_11_05.jpg" class="calibre4"/>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US">Figure 11.5 – Dimmendaal’s hierarchy of Nilo-Saharan subfamilies</p>
<p class="callout-heading">Note</p>
<p class="calibre3">In <em class="italic">Figure 11</em><em class="italic">.5</em>, Dimmendaal’s classification splits languages first by region and then by dialect or region. Note also that some language subfamilies appear to have emerged together (such as Fur and Kunama) without the presence of an intermediary family. This is a more complicated structure than Greenberg’s classification in <em class="italic">Figure 11</em><em class="italic">.4</em>.</p>
<p class="calibre3">As language family hierarchies evolve as languages are studied more systematically over time, it is important to note differences that emerge and their potential causes. This shines light on avenues of further linguistic research and can contribute to future language family hierarchy development.</p>
<p class="calibre3">Now that we know a little bit about Nilo-Saharan languages and their hierarchies, let’s see how network <a id="_idIndexMarker521" class="pcalibre calibre6 pcalibre1"/>science can help us understand <a id="_idIndexMarker522" class="pcalibre calibre6 pcalibre1"/>differences in structure that may be important.</p>
<p class="callout-heading">Note</p>
<p class="calibre3">Our example is very limited in size such that it is easy to visualize results and why our analyses may show differences. Working with full OWL databases or full language hierarchies creates much more nuanced and large networks to study. However, the tools for doing so are the same regardless of network size or complexity. Thus, the analyses we’re about to begin can be applied to much more complicated linguistic data structures where simply visualizing the differences between OWL databases created by different research groups or different language family hierarchies may not be feasible.</p>
<h1 id="_idParaDest-154" class="calibre5"><a id="_idTextAnchor154" class="pcalibre calibre6 pcalibre1"/>Mapping language families</h1>
<p class="calibre3">Let’s return to<a id="_idIndexMarker523" class="pcalibre calibre6 pcalibre1"/> our two examples of Nilo-Saharan language family trees. To examine differences in structure, we’ll need to first create these in NetworkX. For the Dimmendaal family tree, we’ll break with the tree structure slightly to show the interrelationships that exist for some of the child subfamilies. Let’s get started by creating and plotting the Greenberg tree with <code>Script 11.1</code>:</p>
<pre class="source-code">
#load needed packages
import numpy as np
import networkx as nx
import matplotlib.pyplot as plt
#create the Greenberg Nilo-Saharan language family network
G = nx.Graph()
G.add_nodes_from([1, 11])
G.add_edges_from([(1,2),(1,3),(1,4),(1,5),(1,6),(1,7),(7,8),
    (7,9),(7,10),(7,11)])
#plot the Greenberg Nilo-Saharan language family  network
import matplotlib.pyplot as plt
G.nodes[1]['subfamily'] = 'Nilo_Saharan'
G.nodes[2]['subfamily'] = 'Koman'
G.nodes[3]['subfamily'] = 'Saharan'
G.nodes[4]['subfamily'] = 'Songhay'
G.nodes[5]['subfamily'] = 'Fur'
G.nodes[6]['subfamily'] = 'Maban'
G.nodes[7]['subfamily'] = 'Chari_Nile'
G.nodes[8]['subfamily'] = 'Central_Sudanic'
G.nodes[9]['subfamily'] = 'Kunama'
G.nodes[10]['subfamily'] = 'Berta'
G.nodes[11]['subfamily'] = 'Eastern_Sudanic'
labels = nx.get_node_attributes(G, 'subfamily')
nx.draw(G, labels=labels, font_weight='bold')</pre> <p class="calibre3">This script <a id="_idIndexMarker524" class="pcalibre calibre6 pcalibre1"/>produces a similar representation to <em class="italic">Figure 11</em><em class="italic">.4</em>. <em class="italic">Figure 11</em><em class="italic">.6</em> shows the NetworkX plot of the Greenberg Nilo-Saharan language subfamily tree:</p>
<div><div><img alt="Figure 11.6 – A plot of Greenberg’s Nilo-Saharan language subfamily tree" src="img/B21087_11_06.jpg" class="calibre4"/>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US">Figure 11.6 – A plot of Greenberg’s Nilo-Saharan language subfamily tree</p>
<p class="calibre3">Now that we can see Greenberg’s Nilo-Saharan language subfamily tree and its various branches, let’s <a id="_idIndexMarker525" class="pcalibre calibre6 pcalibre1"/>create and plot the Dimmendaal Nilo-Saharan language subfamily tree and its various branches by adding to <code>Script 11.1</code>:</p>
<pre class="source-code">
#create the Dimmendaal Nilo-Saharan language family network
G2 = nx.Graph()
G2.add_nodes_from([1, 9])
G2.add_edges_from([(1,2),(1,3),(2,4),(2,5),(2,6),(5,6),(2,7),
    (2,8),(7,8),(2,9)])
#plot the Dimmendaal Nilo-Saharan language family  network
G2.nodes[1]['subfamily'] = 'Nilo_Saharan'
G2.nodes[2]['subfamily'] = 'Northeastern'
G2.nodes[3]['subfamily'] = 'Central_Sudanic'
G2.nodes[4]['subfamily'] = 'Maban'
G2.nodes[5]['subfamily'] = 'Kunama'
G2.nodes[6]['subfamily'] = 'Fur'
G2.nodes[7]['subfamily'] = 'Saharan'
G2.nodes[8]['subfamily'] = 'Eastern_Sudanic'
G2.nodes[9]['subfamily'] = 'Kuliak'
labels2 = nx.get_node_attributes(G2, 'subfamily')
nx.draw(G2, labels=labels2, font_weight='bold')</pre> <p class="calibre3">This piece of <a id="_idIndexMarker526" class="pcalibre calibre6 pcalibre1"/>the script produces a plot of Dimmendaal’s Nilo-Saharan language subfamily tree, shown in <em class="italic">Figure 11</em><em class="italic">.7</em>:</p>
<div><div><img alt="Figure 11.7 – A plot of Dimmendaal’s Nilo-Saharan language subfamily tree" src="img/B21087_11_07.jpg" class="calibre4"/>
</div>
</div>
<p class="img---caption" lang="en-US" xml:lang="en-US">Figure 11.7 – A plot of Dimmendaal’s Nilo-Saharan language subfamily tree</p>
<p class="calibre3">Note the connections that exist in Dimmendaal’s Nilo-Saharan language subfamily tree between Kunama <a id="_idIndexMarker527" class="pcalibre calibre6 pcalibre1"/>and Fur, as well as Saharan and Eastern Sudanic. These represent related child subfamilies within the Northeastern branch of Nilo-Saharan languages. From <em class="italic">Figures 11.6</em> and <em class="italic">11.7</em>, we can see a similar overall structure, where a lot of shallow, unconnected branches exist. However, Dimmendaal’s Nilo-Saharan language subfamily tree adds an extra layer to the tree and includes the related subfamilies without a common ancestor directly begetting them.</p>
<p class="calibre3">Given these differences, we’d expect some centrality metrics to vary between the two language subfamily trees. We’ll start with some familiar metrics defined in earlier chapters. <code>Script 11.2</code> analyzes the differences in average betweenness centrality across our two language subfamily trees:</p>
<pre class="source-code">
#compare betweenness centrality of language families
gb=nx.betweenness_centrality(G)
print(np.mean(np.array(list(gb.values()))))
gb2=nx.betweenness_centrality(G2)
print(np.mean(np.array(list(gb2.values()))))</pre> <p class="calibre3">We obtain average betweenness centralities of <code>0.131</code> (Greenberg’s) and <code>0.127</code> (Dimmendaal’s), suggesting that bridging properties are similar across these subfamily trees. Let’s assess hubness properties by calculating the average degree centrality by adding to <code>Script 11.2</code>:</p>
<pre class="source-code">
#compare degree centrality of language families
gd=nx.degree_centrality(G)
print(np.mean(np.array(list(gd.values()))))
gd2=nx.degree_centrality(G2)
print(np.mean(np.array(list(gd2.values()))))</pre> <p class="calibre3">We obtain average degree centralities of <code>0.182</code> (Greenberg’s) and <code>0.278</code> (Dimmendaal’s), suggesting that hubness properties vary somewhat across these subfamily trees. Given the added layer of depth connecting other layers of the tree in Dimmendaal’s<a id="_idIndexMarker528" class="pcalibre calibre6 pcalibre1"/> subfamily tree structure, we’d expect to see this difference.</p>
<p class="calibre3">Let’s turn our attention to a centrality metric especially suited for real-world linguistic data (and other ontology or phylogeny network datasets).</p>
<p class="calibre3"><strong class="bold">Subgraph centrality</strong> measures <a id="_idIndexMarker529" class="pcalibre calibre6 pcalibre1"/>how each vertex participates within subgraphs on the network by weighting small subnetworks heavily and larger subnetworks less, using the spectra of the network’s adjacency matrix, and estimating subnetworks through random walks on the network. This blends the properties<a id="_idIndexMarker530" class="pcalibre calibre6 pcalibre1"/> of <strong class="bold">spectral clustering</strong> (seen in <a href="B21087_05.xhtml#_idTextAnchor066" class="pcalibre calibre6 pcalibre1"><em class="italic">Chapter 5</em></a>) and degree centrality to create a hybrid centrality measurement.</p>
<p class="calibre3">We can see this in action on our two subfamily trees by adding to <code>Script 11.2</code>:</p>
<pre class="source-code">
#compare subgraph centrality of language families
gs=nx.subgraph_centrality(G)
print(np.mean(np.array(list(gs.values()))))
gs2=nx.subgraph_centrality(G2)
print(np.mean(np.array(list(gs2.values()))))</pre> <p class="calibre3">This script shows that average subgraph centrality varies between the two subfamily trees, with Greenberg’s average subgraph centrality of <code>2.478</code> and Dimmendaal’s average subgraph centrality of <code>3.276</code>. This suggests that the spectra and degree distributions are different across these hypothesized subfamily trees.</p>
<p class="calibre3">While this example only involves two small ontology structures, these types of analyses can be helpful when working with many ontologies or comparing very large ontologies where visualization is not recommended. For example, we may wish to compare the full Nilo-Saharan language trees hypothesized across many different papers. These trees would include hundreds of vertices and possibly dozens of trees. Tracking the differences visually would be quite a challenge!</p>
<p class="calibre3">When we consider linguistic OWL databases, the scale of our analysis grows even further to hundreds of thousands of vertices to compare across OWL databases. At this scale, it is not possible to visually mine the data in a comparison. Network science provides a way <a id="_idIndexMarker531" class="pcalibre calibre6 pcalibre1"/>to compare the proposed OWL database structures and zero in on differences that exist between those databases to find key differences between the proposed structures. In addition, many of the tools we’ve overviewed scale well to this size of data analysis, providing an efficient way to find these key differences. However, understanding the differences that exist in etymology theory is beyond the scope of this book, and many OWL databases include storage in proprietary systems or are proprietary themselves.</p>
<p class="calibre3">This chapter has focused on the field of linguistics. However, many types of ontologies and phylogenies exist in other disciplines. Zoology and human genomics often use ontologies to study relationships between genes, environments, and gene expression. These ontologies tend to be large, and the curation of knowledge to create these ontologies can result in significant differences or knowledge gaps between different ontologies attempting to capture the same knowledge. Scalable centrality metrics, such as those we overviewed in a simple-to-visualize example, can pinpoint overall and individual branch differences across these complicated ontologies to pinpoint areas of interest and potential caveats of using one open source ontology versus another.</p>
<p class="calibre3">As knowledge in different fields grows, the methodology to compare ontologies will develop further, and network science will play a significant role in the management, understanding, and<a id="_idIndexMarker532" class="pcalibre calibre6 pcalibre1"/> usage of ontologies within different disciplines.</p>
<h1 id="_idParaDest-155" class="calibre5"><a id="_idTextAnchor155" class="pcalibre calibre6 pcalibre1"/>Summary</h1>
<p class="calibre3">In this chapter, we learned how to represent linguistic data in databases and networks. We compared two Nilo-Saharan language subfamily networks using centrality metrics and introduced subgraph centrality as a centrality metric particularly suited to analyzing ontologies and phylogenies. Finally, we expanded our treatment of ontologies to those outside the field of linguistics and discussed how the tools in this chapter can be used to scale these analyses to very large groups of very large ontologies to compare suitability for use in real-world applications.</p>
<p class="calibre3">In the next chapter, we will learn about graph databases, which can be used to store ontology data such as the data discussed in this chapter.</p>
<h1 id="_idParaDest-156" class="calibre5"><a id="_idTextAnchor156" class="pcalibre calibre6 pcalibre1"/>References</h1>
<p class="calibre3">Campbell, L. R. (2018). <em class="italic">How many language families are there in </em><em class="italic">the world?</em></p>
<p class="calibre3">Dimmendaal, G. J. (2019). <em class="italic">A typological perspective on the morphology of Nilo-Saharan languages</em>. In <em class="italic">Oxford Research Encyclopedia </em><em class="italic">of Linguistics</em>.</p>
<p class="calibre3">Dimmendaal, G. J., Ahland, C., Jakobi, A., and Lojenga, C. K. (2019). <em class="italic">Linguistic features and typologies in languages commonly referred to as ‘Nilo-Saharan’</em>. <em class="italic">Cambridge Handbook of African </em><em class="italic">Languages</em>, 326-381.</p>
<p class="calibre3">Done, B., Khatri, P., Done, A., and Draghici, S. (2008). <em class="italic">Predicting novel human gene ontology annotations using semantic analysis</em>. <em class="italic">IEEE/ACM transactions on computational biology and bioinformatics</em>, <em class="italic">7</em>(1), 91-99.</p>
<p class="calibre3">Ehret, C. (2006). <em class="italic">The Nilo-Saharan background of Chadic</em>. <em class="italic">Studies in African Linguistics</em>, <em class="italic">35</em>, 56-66.</p>
<p class="calibre3">Estrada, E., and Rodriguez-Velazquez, J. A. (2005). <em class="italic">Subgraph centrality in complex networks</em>. <em class="italic">Physical Review E</em>, <em class="italic">71</em>(5), 056103.</p>
<p class="calibre3">Gene Ontology Consortium. (2008). <em class="italic">The gene ontology project in 2008</em>. <em class="italic">Nucleic acids research</em>, <em class="italic">36</em>(suppl_1), D440-D444.</p>
<p class="calibre3">Greenberg, J. H. (1957). <em class="italic">Essays </em><em class="italic">in linguistics</em>.</p>
<p class="calibre3">Greenberg, Joseph H. 1963. <em class="italic">The Languages of Africa</em>. Bloomington, IN: Indiana University.</p>
<p class="calibre3">Jakobi, A., and Dimmendaal, G. J. (2022). <em class="italic">Number marking in Karko and Nilo-Saharan</em>. <em class="italic">Number in World’s languages: A Comparative </em><em class="italic">Handbook</em>, 63-106.</p>
<p class="calibre3">McGuinness, D. L., and Van Harmelen, F. (2004). <em class="italic">OWL web ontology language overview</em>. <em class="italic">W3C recommendation</em>, <em class="italic">10</em>(10), 2004.</p>
<p class="calibre3">Oladnabi, M., Omidi, S., Mehrpouya, M., Azadmehr, A., Kazemi-Lomedasht, F., and Yardehnavi, N. (2021). <em class="italic">Venomics and antivenomics data: Current and future perspective</em>. <em class="italic">Archives of Biotechnology and Biomedicine</em>, <em class="italic">5</em>(1), 026-031.</p>
<p class="calibre3">Souag, L. (2022). <em class="italic">How a West African language becomes North African, and vice versa</em>. <em class="italic">Linguistic Typology</em>, <em class="italic">26</em>(2), 283-312.</p>
<p class="calibre3">Subich, V. G. (2020). <em class="italic">STRUCTURAL AND TYPOLOGICAL ANALYSIS OF THE GENEALOGY OF AFRICAN LANGUAGES.</em> Арабистика Евразии, (12), 103-117.</p>
<p class="calibre3">Suryamohan, K., Krishnankutty, S. P., Guillory, J., Jevit, M., Schröder, M. S., Wu, M., ... and Seshagiri, S. (2020). <em class="italic">The Indian cobra reference genome and transcriptome enables comprehensive identification of venom toxins</em>. <em class="italic">Nature Genetics</em>, <em class="italic">52</em>(1), 106-117.</p>
<p class="calibre3">Urban, M. (2021). <em class="italic">The geography and development of language isolates</em>. <em class="italic">Royal Society open science</em>, <em class="italic">8</em>(4), 202232.</p>
<p class="calibre3">Wardhaugh, R. (1972). <em class="italic">Introduction </em><em class="italic">to Linguistics</em>.</p>
<p class="calibre3"><em class="italic">Nigerian Pidgin – 20 useful words and phrases</em> (n.d.): <a href="https://www.britishcouncil.org/voices-magazine/nigerian-pidgin-words-phrases" class="pcalibre calibre6 pcalibre1">https://www.britishcouncil.org/voices-magazine/nigerian-pidgin-words-phrases</a></p>
</div>
</body></html>